onnxruntime/core/graph/graph.cc

```diff
@@ -4213,67 +4250,159 @@ const ONNX_NAMESPACE::GraphProto& Graph::ToGraphProto() {
   return *graph_proto_;
 }
 
-ONNX_NAMESPACE::GraphProto Graph::ToGraphProto() const {
-#if !defined(DISABLE_SPARSE_TENSORS)
-  if (!GraphProtoSyncNeeded() && sparse_tensor_names_.empty()) {
-    return *graph_proto_;
-  }
-#else
-  if (!GraphProtoSyncNeeded()) {
-    return *graph_proto_;
+namespace {
+// The function examines initializer tensor_proto and if the data is in memory it inlines
+// and adds it to the output initializers. Otherwise, it copies the initializer.
+// External data on disk references are preserved.
+Status InlineOrCopyInitializerToGraphProto(const Graph& src_graph, const ONNX_NAMESPACE::TensorProto& initializer,
+                                           ONNX_NAMESPACE::TensorProto& output) {
+  // copy any in-memory external data into raw data
+  if (utils::HasExternalDataInMemory(initializer)) {
+    OrtValue ort_value;
+    ORT_RETURN_IF_NOT(src_graph.GetOrtValueInitializer(initializer.name(), ort_value, /*check_outer_scope=*/false),
+                      "Unable to find OrtValue for initializer with data in memory");
+    ORT_RETURN_IF_NOT(graph_utils::CheckInMemoryDataMatch(initializer, ort_value.Get<Tensor>()),
+                      "Unable to match OrtValue for initializer with data in memory");
+    output = utils::TensorToTensorProto(ort_value.Get<Tensor>(),
+                                        initializer.name(), /* use_tensor_buffer */ false);
+  } else {
+    output = initializer;
```

Comment:
is copying necessary? or can it be avoided if this function takes `initializer` as a non-const reference and optionally updates it?

Reply:
I see that it only needs to be copied at some call sites. in that case, maybe we could have an in-place version and a copying version that calls the in-place one?

also, regarding the name, the "ToGraphProto" part doesn't seem relevant. it is going from a src tensorproto to a dst tensorproto.

Reply:
Good for the name comment.

There is not an in-place version since it is different graphs. I am not sure we need a function for a simple `operator=` invocation.

Reply:
I was referring to call sites like this:
```c++
      if (utils::HasExternalDataInMemory(initializer)) {
        // If the initializer has external data in memory, we need to inline it.
        ORT_RETURN_IF_ERROR(InlineOrCopyInitializer(*this, initializer, initializer));
      }
```

I think we could avoid the copy there. maybe it doesn't matter.

Reply:
There is no copy here since the function would replace inline.